### PR TITLE
Flyin Dockerfile

### PR DESCRIPTION
**Pull request by Future-Outlier**
## Tracking issue
https://github.com/flyteorg/flyte/issues/4284

## Check all the applicable boxes 

- [ ] I updated the documentation accordingly.
- [x] All new and existing tests passed.
- [x] All commits are signed-off.

## Setup Process
Dockerfile
```Dockerfile
ARG PYTHON_VERSION
FROM python:${PYTHON_VERSION}-slim-buster
MAINTAINER Flyte Team <users@flyte.org>
LABEL org.opencontainers.image.source https://github.com/flyteorg/flytekit
WORKDIR /root
ENV PYTHONPATH /root

ARG VERSION
ARG TARGETARCH

# 1. Update the necessary packages for flytekit
# 2. Install code-server
# 3. Download code-server extensions for Python and Jupyter via wget
# 4. Install flytekit and flytekit-flyin with no cache
# 5. Delete apt cache. Reference: https://gist.github.com/marvell/7c812736565928e602c4
# 6. Some packages will create config file under /home by default, so we need to make sure it's writable
# 7. Change the permission of /tmp, so that others can run command on it
RUN apt-get update \
    && apt-get install build-essential git wget -y \
    && mkdir -p /tmp/ \
    && mkdir -p /tmp/code-server \
    && wget --no-check-certificate -O /tmp/code-server/code-server-4.19.0-linux-${TARGETARCH}.tar.gz https://github.com/coder/code-server/releases/download/v4.19.0/code-server-4.19.0-linux-${TARGETARCH}.tar.gz \
    && tar -xzf /tmp/code-server/code-server-4.19.0-linux-${TARGETARCH}.tar.gz -C /tmp/code-server/ \
    && wget --no-check-certificate https://open-vsx.org/api/ms-python/python/2023.20.0/file/ms-python.python-2023.20.0.vsix -P /tmp/code-server \
    && wget --no-check-certificate https://open-vsx.org/api/ms-toolsai/jupyter/2023.9.100/file/ms-toolsai.jupyter-2023.9.100.vsix -P /tmp/code-server \
    && pip install -U git+https://github.com/Future-Outlier/flytekit.git@7470baf933cd8b0da1bc27184ed6dfbb56a25dc2#subdirectory=plugins/flytekit-flyin \
    && apt-get clean autoclean \
    && apt-get autoremove --yes \
    && rm -rf /var/lib/{apt,dpkg,cache,log}/ \
    && useradd -u 1000 flytekit \
    && chown flytekit: /root \
    && chown flytekit: /home \
    && chown -R flytekit:flytekit /tmp \
    && chmod 755 /tmp   

# Set the environment variable for code-server
ENV PATH="/tmp/code-server/code-server-4.19.0-linux-${TARGETARCH}/bin:${PATH}"

USER flytekit

# Install extensions using code-server
# Execution is performed here as code-server configuration depends on the USER setting
# If we install it as ROOT, the config will be stored in /root/.config/code-server/config.yaml
# Now, the config of code-server will be stored in /home/flytekit/.config/code-server/config.yaml
RUN code-server --install-extension /tmp/code-server/ms-python.python-2023.20.0.vsix \
    && code-server --install-extension /tmp/code-server/ms-toolsai.jupyter-2023.9.100.vsix
```

For amd64
```
docker buildx build -t localhost:30000/amd64:amd . -f DockerfileFlyin --platform linux/amd64 --build-arg PYTHON_VERSION=3.9 --no-cache
docker run -it localhost:30000/amd64:amd /bin/bash
code-server --list-extensions
```

For arm64
```
docker buildx build -t localhost:30000/arm64:arm . -f DockerfileFlyin --platform linux/arm64 --build-arg PYTHON_VERSION=3.9 --no-cache
docker run -it localhost:30000/arm64:arm /bin/bash
code-server --list-extensions
```

### Image
amd64
image
image

arm64
image

### 
We skiped the code server download process.
image
We can use the extension by the dockerfile, not download in the runtime.
image

### CODE REVIEW

**Review by ByronHsu — APPROVED** [review]
Thanks!

File: plugins/flytekit-flyin/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,38 @@
+ARG PYTHON_VERSION
+FROM python:${PYTHON_VERSION}-slim-buster
+MAINTAINER Flyte Team <users@flyte.org>
+LABEL org.opencontainers.image.source https://github.com/flyteorg/flytekit
+WORKDIR /root
+ENV PYTHONPATH /root
+
+# Install necessary packages
+RUN apt-get update \
```
Comment (ByronHsu) [inline plugins/flytekit-flyin/Dockerfile:9]: maybe add some comments about what these commands do. it looks a bit overwhelming now

Comment (Future-Outlier) [reply plugins/flytekit-flyin/Dockerfile:9]: no problem!

**Review by ByronHsu — APPROVED** [review]

**Review by pingsutw — APPROVED** [review]

File: .github/workflows/pythonpublish.yml [hunk 2]
```diff
@@ -225,3 +225,45 @@ jobs:
...
+            ghcr.io/${{ github.repository_owner }}/flytekit
+          tags: |
+            flyin-latest
+            flyin-${{ github.sha }}
+            flyin-${{ needs.deploy.outputs.version }}
+      - name: Push Flyin Image to GitHub Registry
+        uses: docker/build-push-action@v2
+        with:
+          context: "./plugins/flytekit-flyin/"
+          platforms: linux/arm64, linux/amd64
```
Comment (eapolinario) [inline .github/workflows/pythonpublish.yml:262]: this image only works on linux/amd64 for now, right?

Comment (pingsutw) [reply .github/workflows/pythonpublish.yml:262]: we should add a build-arg `arch` in the dockerfile, and pass it when building the image

Comment (Future-Outlier) [reply .github/workflows/pythonpublish.yml:262]: I can create a PR for all docker images in flyte-kit, all of them didn't do this

File: plugins/flytekit-flyin/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,46 @@
+ARG PYTHON_VERSION
+FROM python:${PYTHON_VERSION}-slim-buster
+MAINTAINER Flyte Team <users@flyte.org>
+LABEL org.opencontainers.image.source https://github.com/flyteorg/flytekit
+WORKDIR /root
+ENV PYTHONPATH /root
+
+ARG VERSION
+ARG BUILDARCH
```
Comment (eapolinario) [inline plugins/flytekit-flyin/Dockerfile:9]: this should be `TARGETARCH` instead (as per the docker [docs](https://docs.docker.com/engine/reference/builder/?_gl=1*1e06y5d*_ga*MTE3MDg3MDc2NS4xNjkxNDMxNDc3*_ga_XJWPQMJYHQ*MTcwMTI4NjAwNC4yLjAuMTcwMTI4NjAwNC42MC4wLjA.#automatic-platform-args-in-the-global-scope)). We want the architecture that we're building this platform for, not the arch of the node building it.

Comment (Future-Outlier) [reply plugins/flytekit-flyin/Dockerfile:9]: Yes, you are right!
Thank you really much!!!

**Review by pingsutw — APPROVED** [review]

File: .github/workflows/pythonpublish.yml [hunk 4]
```diff
@@ -227,3 +227,45 @@ jobs:
...
+          push: ${{ github.event_name == 'release' }}
+          tags: ${{ steps.flyin-names.outputs.tags }}
+          build-args: |
+            VERSION=${{ needs.deploy.outputs.version }}
```
Comment (eapolinario) [inline .github/workflows/pythonpublish.yml:268]: we missed the python version here (like [this](https://github.com/flyteorg/flytekit/blob/master/.github/workflows/pythonpublish.yml#L142)). Do we need to have multiple versions of python in this image?

Comment (eapolinario) [reply .github/workflows/pythonpublish.yml:268]: This is causing an issue in releases: https://github.com/flyteorg/flytekit/actions/runs/7108516439/workflow#L142

Comment (Future-Outlier) [reply .github/workflows/pythonpublish.yml:268]: no problem, thanks a lot!

**Pull request merged**